text/2017-07-23-relaxed-memory.md

```diff
@@ -0,0 +1,240 @@
+Crossbeam is correct w.r.t. the C11 memory model.  We believe that in order for Crossbeam to indeed
+serve as the library for Rust concurrency, as `java.util.concurrent` does for Java concurrency,
+Crossbeam's correctness should be formally analyzed and hopefully verified.  This RFC aims for
+providing a step towards the goal.
+
+## Non-motivation
+
+Crossbeam is the marriage of EBR and Rust's linear type system with lifetimes.  However, this RFC
+does not aim to clarify how Rust's type system helps to ensure the correctness of Crossbeam.
+
+
+
+# Detailed design
+
+In this section, we present the proposed implementation of Crossbeam in pseudocode, and explain why
+it is correct w.r.t. the C11 memory model.
+
+
+## Proposed implementation
+
+As discussed in Aaron Turon's original [blog post][aturon]
+and [the pull request of a previous RFC][rfc2-pr], Crossbeam guarantees that *every access to an
+object should happen before the object is deallocated*.  In order to provide this guarantee,
+Crossbeam assumes that when a pinned thread marks an object as unlinked, *the thread already removed
+all the references to the object from the memory, and all the other threads do not reintroduce a
+reference to the object in the memory*; and Crossbeam defers the deallocation of an unlinked object
+until all the threads concurrently pinned at the time the object is unlinked, are unpinned.
+
+In order to track the concurrently pinned threads, Crossbeam utilizes the epoch as follows:
+
+- There is a local epoch `th.epoch` for each *pinned* thread `th`, and a global epoch `EPOCH`.
+
+- When an object is deallocated, it is marked with the **current** global epoch, as in the following
+  pseudocode (fences and orderings such as `Relaxed`, `Acquire`, `Release`, `AcqRel`, and `SeqCst`
+  will be explained below):
+
+  ```rust
+  fn unlink(obj) {
+    'L10: atomic::fence(SeqCst);
+    'L11: let epoch = EPOCH.load(Relaxed);
+
+    'L12: // mark `obj` with `epoch`
+  }
+  ```
+
+- Local epoch `th.epoch` is assigned the value of the global epoch `EPOCH` at the time the
+  corresponding thread `th` is pinned, as follows:
+
+  ```rust
+  fn pin(th) {
+    'L20: let e = EPOCH.load(Relaxed);
+    'L21: th.epoch.store(e, Relaxed);
+    'L22: atomic::fence(SeqCst);
+
+    'L23: // now objects marked with `<= e - 2` can be deallocated.
+  }
+  ```
+
+  When a thread is unpinned, its local epoch is set to a sentinel value, as follows:
+
+  ```rust
+  fn unpin(th) {
+    'L30: th.epoch.store(SENTINEL, Release);
+  }
+  ```
+
+- The global epoch `EPOCH` is incremented only when all the local epochs of pinned threads equal to
+  the global epoch, as follows:
+
+  ```rust
+  fn try_advance() {
+    'L40: let G = EPOCH.load(Relaxed);
+    'L41: atomic::fence(SeqCst);
+
+    'L42: for th in threads {
+    'L43:   let e = th.epoch.load(Relaxed);
+    'L44:   if e != SENTINEL && e != G { return; }
+    'L45: }
+    'L46: atomic::fence(Acquire);
+
+    'L47: EPOCH.store(G + 1, Release);
```

Comment:
I think L47 is correct due to the logic around the epochs and insurance that the current thread is also registered. Otherwise it needs to be a CAS right?